### PR TITLE
NEWRELIC-3807 Fix multiple indexes issue

### PR DESCRIPTION
**Pull request by carlossscastro**
This PR addresses https://github.com/newrelic/nri-postgresql/issues/94

This also fixes an issue where only indexes for the first database in the list were reported
Clean up and enhance debug messages

### CODE REVIEW

File: src/metrics/metric_types.go [hunk 1]
```diff
@@ -20,3 +24,61 @@ func (qd QueryDefinition) GetDataModels() interface{} {
 	ptr := reflect.New(reflect.ValueOf(qd.dataModels).Type())
 	return ptr.Interface()
 }
+
+func (qd *QueryDefinition) insertDatabaseNames(databases collection.DatabaseList) *QueryDefinition {
+	databaseList := ""
+	for database := range databases {
+		databaseList += `'` + database + `',`
+	}
+	databaseList = databaseList[0 : len(databaseList)-1]
```
Comment (gsanchezgavier) [inline src/metrics/metric_types.go:33]: perhaps using some of the `Trim` functions on `strings`. It will improve readability and not panic if `len` is 0

File: src/metrics/metric_types.go [hunk 2]
```diff
@@ -20,3 +24,61 @@ func (qd QueryDefinition) GetDataModels() interface{} {
 	ptr := reflect.New(reflect.ValueOf(qd.dataModels).Type())
 	return ptr.Interface()
 }
+
+func (qd *QueryDefinition) insertDatabaseNames(databases collection.DatabaseList) *QueryDefinition {
```
Comment (gsanchezgavier) [inline src/metrics/metric_types.go:28]: is it unclear whether these functions will modify or not the instance calling it. If not we could change the signature not to use the pointer.

File: src/metrics/metric_types.go [hunk 3]
```diff
@@ -20,3 +24,61 @@ func (qd QueryDefinition) GetDataModels() interface{} {
...
+	databaseList := ""
+	for database := range databases {
+		databaseList += `'` + database + `',`
+	}
+	databaseList = databaseList[0 : len(databaseList)-1]
+
+	qd.query = strings.Replace(qd.query, `%DATABASES%`, databaseList, 1)
```
Comment (gsanchezgavier) [inline src/metrics/metric_types.go:35]: here it modifies the calling object but the other similar methods do not

File: src/metrics/metric_types.go [hunk 4]
```diff
@@ -20,3 +24,61 @@ func (qd QueryDefinition) GetDataModels() interface{} {
 	ptr := reflect.New(reflect.ValueOf(qd.dataModels).Type())
 	return ptr.Interface()
 }
+
+func (qd *QueryDefinition) insertDatabaseNames(databases collection.DatabaseList) *QueryDefinition {
+	databaseList := ""
+	for database := range databases {
+		databaseList += `'` + database + `',`
```
Comment (gsanchezgavier) [inline src/metrics/metric_types.go:31]: maybe use sprintf just to be consistent with the other methods?

File: src/metrics/metric_types.go [hunk 5]
```diff
@@ -20,3 +24,61 @@ func (qd QueryDefinition) GetDataModels() interface{} {
...
+	}
+	databaseList = databaseList[0 : len(databaseList)-1]
+
+	qd.query = strings.Replace(qd.query, `%DATABASES%`, databaseList, 1)
+
+	return qd
+}
+
+func (qd *QueryDefinition) insertSchemaTables(schemaList collection.SchemaList) *QueryDefinition {
+	schemaTables := make([]string, 0)
```
Comment (gsanchezgavier) [inline src/metrics/metric_types.go:41]: perhaps also unify this way to create the string of concatenated things with the others `insert` methods?

File: src/metrics/metrics_test.go [hunk 6]
```diff
@@ -331,6 +331,7 @@ func Test_populateIndexMetricsForDatabase(t *testing.T) {
 			"schema1": collection.TableList{
 				"table1": []string{
 					"index1",
+					"index2",
```
Comment (gsanchezgavier) [inline src/metrics/metrics_test.go:334]: could you add an extra database to the mock and test if indexes are collected for both?

File: src/collection/collection.go [hunk 7]
```diff
@@ -159,15 +160,19 @@ func buildSchemaListForDatabase(con *connection.PGSQLConnection) (SchemaList, er
 
 	for index, row := range dataModel {
 		if !row.SchemaName.Valid || !row.TableName.Valid {
-			log.Error("Query responded with a null schema name or table name. Skipping row %d", index)
+			if row.IndexName.Valid {
+				log.Debug("Skipping Index %s. Schema name or Table name null", row.IndexName.String)
```
Comment (paologallinaharbur) [inline src/collection/collection.go:164]: would make sense to print something like:
```
("validSchema: %b, ValidTableName: %b,",  row.SchemaName.Valid, [...])
```

To show what it is complaining about?

File: src/collection/collection.go [hunk 8]
```diff
@@ -159,15 +160,19 @@ func buildSchemaListForDatabase(con *connection.PGSQLConnection) (SchemaList, er
...
+				log.Debug("Query responded with a null schema name or table name. Skipping row %d", index)
+			}
 			continue
 		}
 
 		if _, ok := schemaList[row.SchemaName.String]; !ok {
 			schemaList[row.SchemaName.String] = make(TableList)
 		}
 
-		if _, ok := schemaList[row.TableName.String]; !ok {
```
Comment (paologallinaharbur) [inline src/collection/collection.go:170]: I am not sure I am fully understanding what is going on here 😕 

Before we checked if we had in a schemalist a particular TableList. If not we were creating an entry for that table=SchemaNAme and index=tableName
Now we are doing that if there is not entry for table=SchemaName and index=tableName

Was it a typo/bug?

File: src/metrics/metric_types.go [hunk 9]
```diff
@@ -20,3 +24,61 @@ func (qd QueryDefinition) GetDataModels() interface{} {
...
+
+	newTableDef := &QueryDefinition{
+		dataModels: qd.dataModels,
+		query:      strings.Replace(qd.query, `%SCHEMA_TABLES%`, schemaTablesString, 1),
+	}
+
+	return newTableDef
+}
+
+func (qd *QueryDefinition) insertSchemaTableIndexes(schemaList collection.SchemaList) *QueryDefinition {
```
Comment (paologallinaharbur) [inline src/metrics/metric_types.go:62]: I have some troubles understanding if the code was simply moved or also modified, would make sense to split those two actions in different commits?

File: src/collection/collection.go (hunk 8 above)
Comment (paologallinaharbur) [reply src/collection/collection.go:170]: ahhh it was the bug we are solving 😅

File: src/metrics/metric_types.go (hunk 2 above)
Comment (carlossscastro) [reply src/metrics/metric_types.go:28]: Done

File: src/metrics/metric_types.go (hunk 4 above)
Comment (carlossscastro) [reply src/metrics/metric_types.go:31]: Done 👍

File: src/metrics/metric_types.go (hunk 3 above)
Comment (carlossscastro) [reply src/metrics/metric_types.go:35]: Solved with the fix above of removing pointers

File: src/collection/collection.go (hunk 8 above)
Comment (carlossscastro) [reply src/collection/collection.go:170]: :) Exactly

File: src/metrics/metric_types.go (hunk 9 above)
Comment (carlossscastro) [reply src/metrics/metric_types.go:62]: Absolutely right. Should have kept it in different commits. They were moved and modified :(

File: src/metrics/metric_types.go (hunk 1 above)
Comment (carlossscastro) [reply src/metrics/metric_types.go]: Fixed. Thx

File: src/metrics/metric_types.go (hunk 5 above)
Comment (carlossscastro) [reply src/metrics/metric_types.go:49]: Done also 👍

File: src/collection/collection.go (hunk 7 above)
Comment (carlossscastro) [reply src/collection/collection.go]: Done 👍

File: src/metrics/metrics_test.go (hunk 6 above)
Comment (carlossscastro) [reply src/metrics/metrics_test.go:334]: Done

File: src/collection/collection.go [hunk 10]
```diff
@@ -159,15 +160,19 @@ func buildSchemaListForDatabase(con *connection.PGSQLConnection) (SchemaList, er
 
 	for index, row := range dataModel {
 		if !row.SchemaName.Valid || !row.TableName.Valid {
-			log.Error("Query responded with a null schema name or table name. Skipping row %d", index)
+			if row.IndexName.Valid {
+				log.Debug("Skipping Index %s. Schema name or Table name null. [Schema: %s / Table %s]", row.IndexName.String)
```
Comment (paologallinaharbur) [inline src/collection/collection.go:164]: I see three %s and just one arg, is it a mistake?

Does it make sense to print the boolean (es: row.SchemaName.Valid) instead of a possible null string?

**Review by gsanchezgavier — APPROVED** [review]
thanks for addressing the comments

File: src/collection/collection.go (hunk 10 above)
Comment (carlossscastro) [reply src/collection/collection.go]: 🤦 fixed
